File: R files/transform.R
## This file contains the code for the transformation 
## of the raw bike rental data. it is intended to run in an 
## Azure ML Execute R Script module. By changing
## the following vaiable to false the code will run
## in R or RStudio.
Azure <- FALSE

## If we are in Azure, source the utilities from the zip
## file. The next lines of code read in the dataset, either 
## in Azure ML or from a csv file for testing purposes.
if(Azure){
  source("src/utilities.R")
  BikeShare <- maml.mapInputPort(1)
  BikeShare$dteday <- set.asPOSIXct(BikeShare)
}else{
  BikeShare <- read.csv("BikeSharing.csv", sep = ",", 
                        header = T, stringsAsFactors = F )
  
  ## Select the columns we need
  cols <- c("dteday", "mnth", "hr", "holiday",
            "workingday", "weathersit", "temp",
            "hum", "windspeed", "cnt")
  BikeShare <- BikeShare[, cols]
  
  ## Transform the date-time object
  BikeShare$dteday <- char.toPOSIXct(BikeShare)
  
  ## Normalize the numeric perdictors 
  cols <- c("temp", "hum", "windspeed") 
  BikeShare[, cols] <- scale(BikeShare[, cols])  
}

## Create a new variable to indicate workday
BikeShare$isWorking <- ifelse(BikeShare$workingday & 
                                !BikeShare$holiday, 1, 0)  

## Add a column of the count of months which could 
## help model trend. 
BikeShare <- month.count(BikeShare)

## Create an ordered factor for the day of the week 
## starting with Monday. Note this factor is then 
## converted to an "ordered" numerical value to be
## compatible with Azure ML table data types.
BikeShare$dayWeek <- as.factor(weekdays(BikeShare$dteday))
BikeShare$dayWeek <- as.numeric(ordered(BikeShare$dayWeek, 
                                        levels = c("Monday", 
                                                   "Tuesday", 
                                                   "Wednesday", 
                                                   "Thursday", 
                                                   "Friday", 
                                                   "Saturday", 
                                                   "Sunday")))

## Add a variable with unique values for time of day for working and non-working days.
BikeShare$workTime <- ifelse(BikeShare$isWorking, 
                             BikeShare$hr, 
                             BikeShare$hr + 24) 

## Shift the order of the hour variable so that it is smoothly
## "humped over 24 hours.
BikeShare$xformHr <- ifelse(BikeShare$hr > 4, 
                            BikeShare$hr - 5, 
                            BikeShare$hr + 19)

## Add a variable with unique values for time of day for working and non-working days.
BikeShare$xformWorkHr <- ifelse(BikeShare$isWorking, 
                             BikeShare$xformHr, 
                             BikeShare$xformHr + 24) 

## Output the transformed data frame if in Azure ML.
if(Azure) {
  maml.mapOutputPort('BikeShare')
} 

File: Python files/transform.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 21 12:49:06 2015

@author: Steve Elston
"""


## The main function with a single argument, a Pandas data frame
## from the first input port of the Execute Python Script module.    
def azureml_main(BikeShare):
    import pandas as pd
    from sklearn import preprocessing
    import utilities as ut
    import numpy as np
    import os
 
## If not in the Azure environment, read the data from a csv 
## file for testing purposes.    
    Azure = False  
    if(Azure == False):
        pathName = "C:/Users/Steve/GIT/Quantia-Analytics/AzureML-Regression-Example/Python files"
        fileName = "BikeSharing.csv"
        filePath = os.path.join(pathName, fileName)
        BikeShare = pd.read_csv(filePath)
        
    ## Drop the columns we do not need    
    BikeShare = BikeShare.drop(['instant',
                          'instant',
                          'atemp',
                          'casual',
                          'registered'], 1)  
                          
    ## Normalize the numeric columns
    scale_cols = ['temp', 'hum', 'windspeed']
    arry = BikeShare[scale_cols].as_matrix()
    BikeShare[scale_cols] = preprocessing.scale(arry)                      
                          
    ## Create a new column to indicate if the day is a working day or not.
    work_day = BikeShare['workingday'].as_matrix()
    holiday = BikeShare['holiday'].as_matrix()                       
    BikeShare['isWorking'] = np.where(np.logical_and(work_day == 1, holiday == 0), 1, 0)
    
    ## Compute a new column with the count of months from
    ## the start of the series which can be used to model 
    ## trend
    BikeShare['monthCount'] = ut.mnth_cnt(BikeShare)
    
    ## Shift the order of the hour variable so that it is smoothly
    ## "humped over 24 hours.## Add a column of the count of months which could 
    hr = BikeShare.hr.as_matrix()
    BikeShare['xformHr'] = np.where(hr > 4, hr - 5, hr + 19)

    ## Add a variable with unique values for time of day for working 
    ## and non-working days.
    isWorking = BikeShare['isWorking'].as_matrix()
    BikeShare['xformWorkHr'] = np.where(isWorking, 
                                        BikeShare.xformHr, 
                                        BikeShare.xformHr + 24)
                                        
    BikeShare['dayCount'] = pd.Series(range(BikeShare.shape[0]))/24
                                       
        
    return BikeShare 